compact_str/src/lib.rs

```diff
@@ -1170,12 +1173,98 @@ impl CompactString {
         let mut result = Self::with_capacity(v.len());
         let mut iter = v.iter();
         while let Some(s) = next_char(&mut iter, &mut buf) {
-            // SAFETY: next_char() only returns valid string
+            // SAFETY: next_char() only returns valid strings
             let s = unsafe { std::str::from_utf8_unchecked(s) };
             result.push_str(s);
         }
         result
     }
+
+    #[inline(always)]
+    fn from_utf16x(
+        v: &[u8],
+        from_x: impl Fn(u16) -> u16,
+        from_x_bytes: impl Fn([u8; 2]) -> u16,
+    ) -> Result<Self, Utf16Error> {
+        match unsafe { v.align_to::<u16>() } {
+            (&[], _, &[_]) | (&[_], _, &[]) => {
+                // Input had an odd number of bytes.
+                Err(Utf16Error(()))
+            }
+            (&[], v, _) => {
```

Comment:
The [stdlib](https://doc.rust-lang.org/std/primitive.slice.html#method.align_to) states:
> The method may make the middle slice the greatest length possible for a given type and input slice, but only your algorithm’s performance should depend on that, not its correctness. It is permissible for all of the input data to be returned as the prefix or suffix slice.

So it seems while in the common case `v` should contain all of the data from our original slice, it's not guaranteed.

Could we instead do something like:

```
// Input had an odd number of bytes
if v.len() % 2 != 0 {
  return Err(...)
}

match unsafe { v.align_to::<u16>() } {
  // input is correctly aligned
  (&[], v, &[]) => {
    core::char::decode_utf16(v.iter().copied().map(from_x)
      .collect::<Result<CompactString, _>>()
      .map_err(|_| Utf16Error(()))
  },
  // alignment is off
  _ => {
    // SAFETY: we can always reinterpret a `[u8; 2*N]` slice as `[[u8; 2]; N]`
    let v = unsafe { slice::from_raw_parts(v.as_ptr().cast(), v.len() / 2) };
    std::char::decode_utf16(v.iter().copied().map(from_x_bytes))
        .collect::<Result<CompactString, _>>()
        .map_err(|_| Utf16Error(()))
  }
}
```

Reply:
Ah, I didn't see that. Fixed.